Sprint_4

pavelzver1995 opened the pull request.

Commit:
Добавлены тесты для BooksCollector

Ayantea1 commented on tests.py:
```diff
@@ -18,7 +19,89 @@ def test_add_new_book_add_two_books(self):
+    # чтобы тесты были независимыми в каждом из них создавай отдельный экземпляр класса BooksCollector()
+
+    def test_add_new_book_book_added_with_empty_genre(self):
+        collector = BooksCollector()
```
можно улучшить: общее для всех тестов предусловие можно вынести в фикстуру

Ayantea1 commented on tests.py:
```diff
@@ -18,7 +19,89 @@ def test_add_new_book_add_two_books(self):
+        collector = BooksCollector()
+        collector.add_new_book('Фантастическая книга')
+        collector.set_book_genre('Фантастическая книга', 'Фантастика')
+        assert collector.get_book_genre('Фантастическая книга') == 'Фантастика'
```
Нужно исправить: для get_book_genre должен быть отдельный тест. Тесты не должны дублировать друг друга. В идеале - чтобы в каждом тесте использовался только один проверяемый метод, все остальное можно получить прямым доступом к словарям и спискам

Ayantea1 commented on tests.py:
```diff
@@ -18,7 +19,89 @@ def test_add_new_book_add_two_books(self):
+        collector.set_book_genre('Комедия1', 'Комедии')
+        
+        horror_books = collector.get_books_with_specific_genre('Ужасы')
+        assert len(horror_books) == 2
```
можно улучшить: достаточно одного ассерта, чтобы проверить вид списка

Ayantea1 commented on tests.py:
```diff
@@ -18,7 +19,89 @@ def test_add_new_book_add_two_books(self):
+        collector.add_book_in_favorites('Избранная книга')
+        assert 'Избранная книга' in collector.get_list_of_favorites_books()
+        
+        collector.delete_book_from_favorites('Избранная книга')
```
Нужно исправить: каждый тест должен проверять только один метод

Commit:
Добавлена фикстура collector - создает отдельный экземпляр BooksCollector для каждого теста

Разделены тесты по методам - каждый тест проверяет только один метод

Убраны лишние вызовы методов - для проверки используются прямые обращения к атрибутам (books_genre, favorites)

Добавлены отдельные тесты для всех методов:

test_get_book_genre_returns_correct_genre

test_get_books_genre_returns_current_dict

test_delete_book_from_favorites_existing_book

test_get_list_of_favorites_books_returns_current_list

Упрощены проверки - используется один ассерт там, где это достаточно

Добавлены тесты для граничных случаев:

Несуществующие книги

Несуществующие жанры

Удаление несуществующих книг из избранного

Теперь каждый тест независим и проверяет только одну функциональность

Ayantea1 commented on tests.py:
```diff
@@ -5,103 +5,116 @@
 class TestBooksCollector:
 
+    # фикстура для создания экземпляра коллектора
+    @pytest.fixture
```
можно улучшить: фикстуру лучше хранить в отдельном модуле
